torch/csrc/autograd/python_function.cpp

```diff
@@ -487,7 +487,7 @@ std::pair<UnpackedInput, InputFlags> unpack_input(PyObject *args) {
   }
 
   flags.is_executable = GradMode::is_enabled() && any_variable_requires_grad(unpacked.input_vars);
-  flags.next_edges = collect_next_edges(unpacked.input_vars);
+  flags.next_edges = (flags.is_executable ? collect_next_edges(unpacked.input_vars) : edge_list());
```

Comment:
nit: this would actually change the current behavior as we don't build the edges anymore when no input requires grad?
It is better in terms of speed but is changing the current behavior. I am not familiar enough with this function's callers to know if this change is ok?

Reply:
You're definitely right that the behavior is different now. I went back and forth on whether to use the `is_executable` flag here instead of `GradMode::is_enabled()`, but convinced myself through tracing of the logic that the edges will be accessed based on the `is_executable` flag, certainly in `custom_function.h`, and almost certainly here in `python_function.cpp`.

Do you think the existing tests will catch this if this assumption is wrong? I'm also okay with changing the check back to `GradMode::enabled()` to be extra safe.

Reply:
> Do you think the existing tests will catch this if this assumption is wrong?

We only have "end user" tests for these things. So if the test pass it should be good. But not 100% sure :p 

If you went and checked the callsite, that is good enough for me!